### ISSUE TITLE
Create channel Api call fails if Api is logged in with username and password.

### ISSUE THREAD

**Issue by nikkilocke** [mattermost/mattermost-server#11590]
#### Summary
Create channel Api call fails if Api is logged in with username and password. 

#### Steps to reproduce
If I use the api to login using username and password, any attempt to create a channel fails with `401 Unauthorized`, and the json: 

		  "id": "api.context.session_expired.app_error",
		  "message": "Invalid or expired session, please login again.",
		  "detailed_error": "",
		  "request_id": "nfrr7jrezfg55ndy67ba3oqchy",
		  "status_code": 401,

If I use the api to login a trusted application with OAuth, it works fine.

#### Expected behavior
Create channel request succeeds

#### Observed behavior (that appears unintentional) 
Create channel call fails. 

#### Possible fixes
If you can, link to the line of code that might be responsible for the problem

**Comment by amyblais** [comment]
@maruTA-bis5 @hanzei Do you know if this is a bug?

**Comment by amyblais** [comment]
@nikkilocke Would you be open to sharing any additional information for us to investigate this report?

**Comment by rvillablanca** [comment]
I could not reproduce this issue. After a success authentication I was able to create any channels.

@nikkilocke Maybe you forgot adding the token as bearear authentication when doing the create channel request

**Comment by nikkilocke** [comment]
Here is a log of a session where I logged in, received a token, and used it. The first time I tried to POST, it returned Unauthorized. Please can you explain what I am doing wrong?

		Sent -> Method: POST, RequestUri: 'http://ubuntu:8065/api/v4/users/login', Version: 2.0, Content: System.Net.Http.StringContent, Headers:
		{
		  Accept: application/json
		  Accept: text/html
		  Accept: */*
		  User-Agent: BaseCampGateway
		  Content-Type: application/json; charset=utf-8
		}:{
		  "login_id": "basecamp_bot",
		  "password": "mypassword"
		}
		Received -> StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
		{
		  Set-Cookie: MMAUTHTOKEN=5okk8z4uyjf1dqrobpfwuaosxc; Path=/; Expires=Sun, 12 Jan 2020 13:54:10 GMT; Max-Age=15552000; HttpOnly
		  Set-Cookie: MMUSERID=f9gsc3saobgn3gffxkxku5cakc; Path=/; Expires=Sun, 12 Jan 2020 13:54:10 GMT; Max-Age=15552000
		  Set-Cookie: MMCSRF=ne7yuahqqj817n9i1wuuxy5o6e; Path=/; Expires=Sun, 12 Jan 2020 13:54:10 GMT; Max-Age=15552000
		  Token: 5okk8z4uyjf1dqrobpfwuaosxc
		  X-Request-ID: 3hhk4qs6obby3fj4b74h4j1bmc
		  X-Version-Id: 5.11.0.5.11.0.aa195431e463287989705167befb35df.false
		  Date: Tue, 16 Jul 2019 13:54:10 GMT
		  Content-Type: application/json
		  Content-Length: 651
		}
		Received Data -> {
		  "id": "f9gsc3saobgn3gffxkxku5cakc",
		  "create_at": 1563285238107,
		  "update_at": 1563285238342,
		  "delete_at": 0,
		  "username": "basecamp_bot",
		  "auth_data": "",
		  "auth_service": "",
		  "email": "nikkilocke@myemail.com",
		  "nickname": "",
		  "first_name": "",
		  "last_name": "",
		  "position": "",
		  "roles": "system_user",
		  "allow_marketing": true,
		  "notify_props": {
			"channel": "true",
			"comments": "never",
			"desktop": "mention",
			"desktop_sound": "true",
			"email": "true",
			"first_name": "false",
			"mention_keys": "basecamp_bot,@basecamp_bot",
			"push": "mention",
			"push_status": "away"
		  },
		  "last_password_update": 1563285238107,
		  "locale": "en",
		  "timezone": {
			"automaticTimezone": "",
			"manualTimezone": "",
			"useAutomaticTimezone": "true"
		  },
		  "MetaData": {
			"Uri": "http://ubuntu:8065/api/v4/users/login"
		  }
		}

		Sent -> Method: GET, RequestUri: 'http://ubuntu:8065/api/v4/teams/k9x58mi7ttd5fgjho4f9tb51hw', Version: 2.0, Content: <null>, Headers:
		{
		  Authorization: Bearer 5okk8z4uyjf1dqrobpfwuaosxc
		  Accept: application/json
		  Accept: text/html
		  Accept: */*
		  User-Agent: BaseCampGateway
		}:
		Received -> StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
		{
		  X-Request-ID: wxw5pntrxf84pbmh6t5t9imb3r
		  X-Version-Id: 5.11.0.5.11.0.aa195431e463287989705167befb35df.false
		  Date: Tue, 16 Jul 2019 13:54:12 GMT
		  Content-Type: application/json
		  Expires: 0
		  Content-Length: 323
		}
		Received Data -> {
		  "id": "k9x58mi7ttd5fgjho4f9tb51hw",
		  "create_at": 1560267803965,
		  "update_at": 1560952877555,
		  "delete_at": 0,
		  "display_name": "UK",
		  "name": "xr-uk",
		  "description": "",
		  "email": "",
		  "type": "O",
		  "company_name": "",
		  "allowed_domains": "",
		  "invite_id": "rpedbk1q6ffg3kzbhuekgroesa",
		  "allow_open_invite": false,
		  "scheme_id": null,
		  "group_constrained": null,
		  "MetaData": {
			"Uri": "http://ubuntu:8065/api/v4/teams/k9x58mi7ttd5fgjho4f9tb51hw"
		  }
		}

		Sent -> Method: POST, RequestUri: 'http://ubuntu:8065/api/v4/teams/k9x58mi7ttd5fgjho4f9tb51hw/members', Version: 2.0, Content: System.Net.Http.StringContent, Headers:
		{
		  Authorization: Bearer 5okk8z4uyjf1dqrobpfwuaosxc
		  Accept: application/json
		  Accept: text/html
		  Accept: */*
		  User-Agent: BaseCampGateway
		  Content-Type: application/json; charset=utf-8
		}:{
		  "team_id": "k9x58mi7ttd5fgjho4f9tb51hw",
		  "user_id": "f9gsc3saobgn3gffxkxku5cakc"
		}
		Received -> StatusCode: 401, ReasonPhrase: 'Unauthorized', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
		{
		  X-Request-ID: 5ejrhrd357dzig9krpq9ogs97h
		  X-Version-Id: 5.11.0.5.11.0.aa195431e463287989705167befb35df.false
		  Date: Tue, 16 Jul 2019 13:55:23 GMT
		  Content-Type: application/json
		  Content-Length: 186
		}
		Received Data -> {
		  "id": "api.context.session_expired.app_error",
		  "message": "Invalid or expired session, please login again.",
		  "detailed_error": "",
		  "request_id": "5ejrhrd357dzig9krpq9ogs97h",
		  "status_code": 401,
		  "MetaData": {
			"Error": {
			  "id": "api.context.session_expired.app_error",
			  "message": "Invalid or expired session, please login again.",
			  "detailed_error": "",
			  "request_id": "5ejrhrd357dzig9krpq9ogs97h",
			  "status_code": 401
			},
			"Uri": "http://ubuntu:8065/api/v4/teams/k9x58mi7ttd5fgjho4f9tb51hw/members"
		  }
		}

Am I supposed to send back the cookies as well as the Authorization: Bearer?

**Comment by amyblais** [comment]
Hi @nikkilocke, I asked our team about and got this response from a community member: 

You are using ``POST /api/v4/teams/{team_id}/members (Add user to team)``. To create a channel, you would need to use ``POST /api/v4/channels``.

**Comment by nikkilocke** [comment]
Thanks for asking. Unfortunately I have obviously not explained the problem very well. In the case of the log I sent you, I was trying to add a user to the team, which also fails in exactly the same way. I suspect all POSTS might be failing if I log on to the Api with username and password instead of using oauth. Perhaps you could tell me what information you need from me to get this problem sorted out? It is preventing me from continuing with my project, moving about 8-10 thousand users of our campaigning group from Basecamp to Mattermost and Discourse. Regards Nikki On 23 Jul 2019 7:29 pm, amyblais <notifications@github.com> wrote:Hi @nikkilocke, I asked our team about and got this response from a community member:
You are using POST /api/v4/teams/{team_id}/members (Add user to team). To create a channel, you would need to use POST /api/v4/channels.

**Comment by nikkilocke** [comment]
Herewith a new log, concentrating on the create channel api.
[buglog.txt](https://github.com/mattermost/mattermost-server/files/3426665/buglog.txt)

**Comment by maruTA-bis5** [comment]
Could you share the server log (mattermost.log) that includes `debug` level?

**Comment by nikkilocke** [comment]
Sure - here.
[mattermost.log](https://github.com/mattermost/mattermost-server/files/3430513/mattermost.log)
[bug.log](https://github.com/mattermost/mattermost-server/files/3430514/bug.log)

**Comment by nikkilocke** [comment]
Aha! By looking through your source code, and doing a lot of guessing, i have solved the problem. Logging in with a username and password returns an `MMCSRF` cookie. This cookie has to be included in an `X-CSRF-Token` header on all subsequent requests (except GET, which is allowed, for some reason).

Can I strongly suggest that you document this in the Mattermost API docs?

**Comment by maruTA-bis5** [comment]
According to server source code, CSRF check fires when `tokenLocation == app.TokenLocationCookie`.
This means you are sending `MMAUTHTOKEN` cookie to server (so `Authorization` header is ignored).
https://github.com/mattermost/mattermost-server/blob/692d1be91eafb64328ce21bcd18671288eb5139b/web/handlers.go#L126-L128

**Comment by nikkilocke** [comment]
Thanks very much for that - C# HttpClient tracks and sends cookies even if you don't ask it to, which I wasn't aware of. I still suggest you document this behaviour to help the next person.

**Comment by jasonblais** [comment]
Glad your issue was resolved @nikkilocke and thank you @rvillablanca @maruTA-bis5 so much for jumping in to help with the issue.

@nikkilocke @rvillablanca @maruTA-bis5 Wondering if any of you were open to helping with a pull request to make this more clear in the documentation?

**Comment by amyblais** [comment]
Opened a Help Wanted issue in API repo for anyone to document this helpful information: https://github.com/mattermost/mattermost-api-reference/issues/467.

**Comment by Niceplace** [comment]
Adding a comment here if anyone else experiences this issue. I am using Insomnia REST client and I was chaining a LOGIN request do a DELETE request on /teams. I kept getting that issue:

```
{
  "id": "api.context.session_expired.app_error",
  "message": "Invalid or expired session, please login again.",
  "detailed_error": "token=yoyzcjj1bpfwx8odfm8twkkakc Appears to be a CSRF attempt",
  "request_id": "8h5tq5zdxfgj58eiwyk8tja6qc",
  "status_code": 401
}
```

I had to go in the request settings for the LOGIN request and disable 'Send cookies automatically' and 'Store cookies automatically'. After this, referencing the Bearer token of the LOGIN response in the DELETE request worked without giving the CSRF related error.

Thanks for the pointers folks !
